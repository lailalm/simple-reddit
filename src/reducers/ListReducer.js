import {
  GENERATE_TOPIC,
  UPVOTE_TOPIC,
  DOWNVOTE_TOPIC,
} from 'utils/ActionTypes';

const initialState = {
  data: [
    {
      id: '1',
      title: 'First Post',
      upvotes: 5,
      downvotes: 3,
    },
    {
      id: '2',
      title: 'Second Post',
      upvotes: 3,
      downvotes: 3,
    },
    {
      id: '3',
      title: 'Third Post',
      upvotes: 0,
      downvotes: 1,
    },
  ],
  sortOptions: 'DECREASING_UPVOTE',
};

export default (state = initialState, action) => {
  const { type, payload } = action;
  let data;
  switch (type) {
    case UPVOTE_TOPIC:
      data = state.data.map((item) => {
        if (item.id !== payload.id) {
          return item;
        }
        return {
          ...item,
          upvotes: item.upvotes + 1,
        };
      });
      return {
        ...state,
        data,
      };
    case DOWNVOTE_TOPIC:
      data = state.data.map((item) => {
        if (item.id !== payload.id) {
          return item;
        }
        return {
          ...item,
          downvotes: item.downvotes + 1,
        };
      });
      return {
        ...state,
        data,
      };
    case GENERATE_TOPIC:
      return {
        ...state,
        data: {
          ...state.data,
          payload,
        },
      };
    default:
      return state;
  }
};
