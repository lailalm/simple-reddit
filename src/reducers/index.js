import { combineReducers } from 'redux';
import ListReducer from './ListReducer';
import Routes from './Routes';

export default combineReducers({
  listReducer: ListReducer,
  Routes,
});
