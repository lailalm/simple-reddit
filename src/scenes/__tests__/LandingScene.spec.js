import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import { shallow, mount } from 'enzyme';

import ConnectedLandingScene, { LandingScene } from 'scenes/LandingScene';

import {
  upvoteTopic,
  downvoteTopic,
} from 'actions/TopicAction';

import {
  UPVOTE_TOPIC,
  DOWNVOTE_TOPIC,
} from 'utils/ActionTypes';

import reducers from 'reducers/';

const dummyOneData = {
  data: [{
    id: '1',
    title: 'First Post',
    upvotes: 5,
    downvotes: 3,
  }],
  sortOptions: 'DECREASING_UPVOTE',
};

// Snapshot for LandingScene React Component
describe('>>> LandingScene --- Snapshot', () => {
  it('+++ capturing Snapshot of LandingScene', () => {
    const renderedValue = renderer.create(<LandingScene list={dummyOneData} />).toJSON();
    expect(renderedValue).toMatchSnapshot();
  });
});

describe('>>> LandingScene --- Shallow Render', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<LandingScene list={dummyOneData} />);
  });

  it('+++ render the dummy component', () => {
    expect(wrapper.length).toEqual(1);
  });
});

describe('>>> LandingScene --- React-Redux (Shallow + passing the {store} directly)', () => {
  const initialState = {
    listReducer: dummyOneData,
  };
  const mockStore = configureStore();
  let store;
  let container;

  beforeEach(() => {
    store = mockStore(initialState);
    container = shallow(<ConnectedLandingScene store={store} />);
  });

  it('+++ render the connected component', () => {
    expect(container.length).toEqual(1);
  });

  it('+++ check the props match with initialState', () => {
    expect(container.prop('list')).toEqual(initialState.listReducer);
  });
});

describe('>>> LandingScene --- React-Redux (Mount + wrapping in <Provider>)', () => {
  const origConsole = console.error;
  const initialState = {
    listReducer: dummyOneData,
  };
  const mockStore = configureStore();
  let store;
  let wrapper;

  beforeEach(() => {
    console.error = () => {};
    store = mockStore(initialState);
    wrapper = mount(<Provider store={store} ><ConnectedLandingScene /></Provider>);
  });

  // HACK: There isn't any react native adapter, so I'm using the react adapter.
  //       This will temporarily silence propType warnings
  afterEach(() => {
    console.error = origConsole;
  });

  it('+++ render the connected component', () => {
    expect(wrapper.find(ConnectedLandingScene).length).toEqual(1);
  });

  it('+++ check prop matches with initialState', () => {
    expect(wrapper.find(LandingScene).prop('list')).toEqual(initialState.listReducer);
  });

  it('+++ check action on dispatching ', () => {
    let action;
    store.dispatch(upvoteTopic(1));
    store.dispatch(downvoteTopic(1));
    action = store.getActions();
    expect(action[0].type).toBe(UPVOTE_TOPIC);
    expect(action[1].type).toBe(DOWNVOTE_TOPIC);
  });
});

describe('>>> LandingScene --- React-Redux (actual Store + reducers)', () => {
  const origConsole = console.error;
  let store;
  let wrapper;

  beforeEach(() => {
    console.error = () => {};
    store = createStore(reducers);
    wrapper = mount(<Provider store={store}><ConnectedLandingScene /></Provider>);
  });

  // HACK: There isn't any react native adapter, so I'm using the react adapter.
  //       This will temporarily silence propType warnings
  afterEach(() => {
    console.error = origConsole;
  });

  it('+++ check Prop matches with initialState', () => {
    expect(wrapper.find(LandingScene).prop('list')).toEqual(expect.objectContaining({
      data: expect.arrayContaining(dummyOneData.data),
      sortOptions: expect.any(String),
    }));
  });
});
