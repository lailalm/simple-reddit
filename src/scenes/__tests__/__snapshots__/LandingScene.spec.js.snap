// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> LandingScene --- Snapshot +++ capturing Snapshot of LandingScene 1`] = `
<RCTScrollView>
  <View>
    <View
      style={
        Array [
          Object {
            "marginLeft": 15,
            "marginRight": 15,
          },
          undefined,
          undefined,
          undefined,
        ]
      }
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={null}
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <View
          pointerEvents="box-only"
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#9E9E9E",
                "flexDirection": "row",
                "justifyContent": "center",
                "padding": 19,
              },
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              Object {
                "padding": 12,
              },
              Object {
                "borderRadius": 66.5,
                "paddingHorizontal": 18,
              },
              undefined,
              undefined,
              Object {
                "backgroundColor": "#67697C",
                "marginTop": 20,
              },
              undefined,
              undefined,
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={false}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "white",
                  "fontSize": 18,
                },
                Array [
                  Object {
                    "marginRight": 10,
                  },
                  undefined,
                ],
                Object {
                  "fontFamily": "Material Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
              ]
            }
          >
            
          </Text>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {},
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                Array [
                  Object {
                    "color": "white",
                    "fontSize": 20,
                  },
                  undefined,
                  Object {
                    "fontSize": 17.5,
                  },
                  undefined,
                  Object {
                    "textAlign": "center",
                  },
                  undefined,
                  undefined,
                  undefined,
                ],
              ]
            }
          >
            Create New Post
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "white",
            "borderColor": "#e1e8ee",
            "borderWidth": 1,
            "margin": 15,
            "marginBottom": 0,
            "padding": 15,
            "shadowColor": "rgba(0,0,0, .2)",
            "shadowOffset": Object {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 1,
            "shadowRadius": 1,
          },
          undefined,
          undefined,
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
            },
            undefined,
            undefined,
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {},
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
            ]
          }
        >
          First Post
        </Text>
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "row",
              "marginTop": 10,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 2,
                "flexDirection": "row",
              }
            }
          >
            <View
              accessibilityComponentType={undefined}
              accessibilityLabel={undefined}
              accessibilityTraits={undefined}
              accessible={true}
              hasTVPreferredFocus={undefined}
              hitSlop={undefined}
              isTVSelectable={true}
              nativeID={undefined}
              onLayout={undefined}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Array [
                  Object {
                    "margin": 7,
                  },
                  Object {
                    "borderRadius": 24,
                    "height": 44,
                    "width": 44,
                  },
                  false,
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#51A3A3",
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
              testID={undefined}
              tvParallaxProperties={undefined}
            >
              <Text
                accessible={true}
                allowFontScaling={false}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "color": "white",
                      "fontSize": 20,
                    },
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                      },
                      undefined,
                    ],
                    Object {
                      "fontFamily": "fontcustom",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                  ]
                }
              >
                
              </Text>
            </View>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {},
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                ]
              }
            >
              5
               upvotes
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 2,
                "flexDirection": "row",
              }
            }
          >
            <View
              accessibilityComponentType={undefined}
              accessibilityLabel={undefined}
              accessibilityTraits={undefined}
              accessible={true}
              hasTVPreferredFocus={undefined}
              hitSlop={undefined}
              isTVSelectable={true}
              nativeID={undefined}
              onLayout={undefined}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Array [
                  Object {
                    "margin": 7,
                  },
                  Object {
                    "borderRadius": 24,
                    "height": 44,
                    "width": 44,
                  },
                  false,
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#DD7373",
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
              testID={undefined}
              tvParallaxProperties={undefined}
            >
              <Text
                accessible={true}
                allowFontScaling={false}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "color": "white",
                      "fontSize": 20,
                    },
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                      },
                      undefined,
                    ],
                    Object {
                      "fontFamily": "fontcustom",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                  ]
                }
              >
                
              </Text>
            </View>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {},
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                ]
              }
            >
              3
               downvotes
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
